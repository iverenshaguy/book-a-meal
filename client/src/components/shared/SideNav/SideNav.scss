@import '../../../../public/scss/_variables';
.sidenav {
  box-shadow: 3px 0 10px 1px rgba(0, 0, 0, 0.05);
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  z-index: 11;
  margin-top: 0;
}

@media (max-width: 767.98px) {
  .sidenav {
    position: relative;
    &:before {
      content: "";
      display: none;
      background: rgba(0, 0, 0, 0.6);
      position: fixed;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      z-index: 11;
    }
    &:target,
    &.show {
      &:before {
        display: block;
      }
      .sidenav-content {
        -webkit-transform: translate(0, 0);
        transform: translate(0, 0);
      }
    }
    .sidenav-content {
      z-index: 12;
      position: fixed;
      top: 0;
      left: 0;
      -webkit-transform: translate(-100%, 0);
      transform: translate(-100%, 0);
      transition: -webkit-transform 0.5s ease-out;
      transition: transform 0.5s ease-out;
    }
  }
}

.sidenav-content {
  height: 100%;
  width: 250px;
  background-color: $white;
  overflow-x: hidden;
  .sidenav-header {
    display: flex;
    flex-direction: column;
    justify-content: center;
    /* background-color: $darkgrey; */
    height: 300px;
    padding-bottom: 20px;
    color: $dark;
    .sidenav-title {
      color: $dark;
      flex-grow: 11;
      font-size: 1.3em;
      font-weight: bold;
      font-family: $heading-font-stack;
      font-display: swap;
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      h3 {
        margin-top: -1em;
      }
      p {
        margin: 0.7em 0;
      }
      span a,
      span .link-btn {
        color: $dark;
        font-size: 0.7em;
        &:hover,
        &:focus,
        &.active {
          transition: color 0.5s;
          color: $primary-color;
        }
      }
    }
    .close {
      flex-grow: 1;
      text-align: right;
      display: flex;
      flex-direction: row-reverse;
      a,
      .link-btn {
        font-size: 1.5em;
        font-weight: bold;
        display: block;
        color: $dark;
        margin: 0.7em 0.7em 0.7em 0;
      }
    }
  }
  .menu-item {
    padding: 15px 15px 15px 32px;
    width: 100%;
    text-align: justify;
    text-decoration: none;
    font-size: 1.1em;
    display: block;
    font-family: $font-stack;
    font-display: swap;
    &:hover,
    &:focus {
      background-color: $lightgrey;
      transition: background-color 0.5s;
    }
    &.active {
      background-color: $lightgrey;
      transition: background-color 0.5s;
      border-left: 3px solid $primary-color;
    }
  }
}

@media (max-height: 450px) {
  .sidenav a,
  .sidenav .link-btn {
    font-size: 18px;
  }
}

@media (min-width: 768px) {
  .sidenav .menu-item {
    padding: 15px 15px 15px 32px;
    text-decoration: none;
    font-size: 1.1em;
    display: block;
  }
  .sidenav-content .sidenav-header {
    padding-bottom: 0;
  }
}