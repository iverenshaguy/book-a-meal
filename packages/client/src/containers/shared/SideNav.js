import { connect } from 'react-redux';
import { push } from 'connected-react-router';
import SideNav from '../../components/shared/SideNav';
import { toggleSideNav } from '../../actions/ui';

const mapStateToProps = state => ({
  open: state.ui.sideNav.open
});

export default connect(mapStateToProps, { toggleSideNav, push })(SideNav);
