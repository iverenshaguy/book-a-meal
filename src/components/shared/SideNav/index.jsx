import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import LinkBtn from '../Link';
import CloseIcon from '../CloseIcon';
import { userPropTypes } from '../../../helpers/proptypes';
import './SideNav.scss';

/**
 * @exports
 * @function SideNav
 * @returns {JSX} SideNav
 */
class SideNav extends Component {
  static propTypes = {
    ...userPropTypes,
    logout: PropTypes.func.isRequired,
    active: PropTypes.string.isRequired
  };

  /**
   * @memberof SideNav
   * @param {string} location
   * @returns {void}
   */
  handleLinkClick = (location) => {
    this.props.toggleSideNav();
    this.props.push(location);
  }

  /**
   * @memberof SideNav
   * @returns {void}
   */
  handleToggleSideNavClick = () => this.props.toggleSideNav()

  /**
   * @memberof SideNav
   * @returns {JSX} NavLinks
  */
  renderNavLinks = () => {
    const { active } = this.props;

    return (
      <div className="sidenav-body">
        <LinkBtn clickHandler={() => this.handleLinkClick('/')} className={`menu-item ${active === 'dashboard' && 'active'}`}>Dashboard</LinkBtn>
        <LinkBtn clickHandler={() => this.handleLinkClick('/meals')} className={`menu-item ${active === 'meals' && 'active'}`}>Meals</LinkBtn>
        <LinkBtn clickHandler={() => this.handleLinkClick('/menu')} className={`menu-item ${active === 'menu' && 'active'}`}>Menu</LinkBtn>
        <LinkBtn clickHandler={() => this.handleLinkClick('/orders')} className={`menu-item ${active === 'orders' && 'active'}`}>Orders</LinkBtn>
      </div>
    );
  }

  /**
   * @memberof SideNav
   * @returns {JSX} SideNav Component
  */
  render() {
    const { user, logout, open } = this.props;

    return (
      <div className={`sidenav ${open ? 'show' : null}`} id="sidenav">
        <div className="sidenav-content">
          <div className="sidenav-header">
            <CloseIcon divClass="d-none-md" clickHandler={this.handleToggleSideNavClick} />
            <div className="sidenav-title">
              <h3 className="d-flex-md"><Link href="/" to="/">BOOK-A-MEAL</Link></h3>
              <div className="username-circle"><p>{user.businessName.substring(0, 1)}</p></div>
              <p>{user.businessName}</p>
              <span><LinkBtn clickHandler={logout}>Logout</LinkBtn></span>
            </div>
          </div>
          {this.renderNavLinks()}
        </div>
      </div>
    );
  }
}

export default SideNav;
