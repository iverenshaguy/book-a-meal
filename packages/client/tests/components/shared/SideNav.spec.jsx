import React from 'react';
import SideNav from '../../../src/components/shared/SideNav';
import { caterer } from '../../setup/mockData';

describe('SideNav', () => {
  it('should render SideNav component correctly', () => {
    const shallowWrapper = shallow(<SideNav user={caterer} open={false} logout={jest.fn()} active="dashboard" />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render correctly when meals tab is active', () => {
    const shallowWrapper = shallow(<SideNav user={caterer} open={false} logout={jest.fn()} active="meals" />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render correctly when menu tab is active', () => {
    const shallowWrapper = shallow(<SideNav user={caterer} open={false} logout={jest.fn()} active="menu" />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render correctly when orders tab is active', () => {
    const shallowWrapper = shallow(<SideNav user={caterer} open={false} logout={jest.fn()} active="orders" />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should toggle sidenav and handle link click when dashboard link is clicked', () => {
    const toggleSideNavMock = jest.fn();
    const pushMock = jest.fn();

    const wrapper = mount(<SideNav user={caterer} logout={jest.fn()} open active="orders" toggleSideNav={toggleSideNavMock} push={pushMock} />, rrcMock.get());

    wrapper.find('div.sidenav-body>Link>button').at(0).simulate('click');
    expect(toggleSideNavMock).toHaveBeenCalled();
    expect(pushMock).toHaveBeenCalledWith('/');
  });

  it('should toggle sidenav and handle link click when meals link is clicked', () => {
    const toggleSideNavMock = jest.fn();
    const pushMock = jest.fn();

    const wrapper = mount(<SideNav user={caterer} logout={jest.fn()} open active="orders" toggleSideNav={toggleSideNavMock} push={pushMock} />, rrcMock.get());

    wrapper.find('div.sidenav-body>Link>button').at(1).simulate('click');
    expect(toggleSideNavMock).toHaveBeenCalled();
    expect(pushMock).toHaveBeenCalledWith('/meals');
  });

  it('should toggle sidenav and handle link click when menu link is clicked', () => {
    const toggleSideNavMock = jest.fn();
    const pushMock = jest.fn();

    const wrapper = mount(<SideNav user={caterer} logout={jest.fn()} open active="orders" toggleSideNav={toggleSideNavMock} push={pushMock} />, rrcMock.get());

    wrapper.find('div.sidenav-body>Link>button').at(2).simulate('click');
    expect(toggleSideNavMock).toHaveBeenCalled();
    expect(pushMock).toHaveBeenCalledWith('/menu');
  });

  it('should toggle sidenav and handle link click when orders link is clicked', () => {
    const toggleSideNavMock = jest.fn();
    const pushMock = jest.fn();

    const wrapper = mount(<SideNav user={caterer} logout={jest.fn()} open active="menu" toggleSideNav={toggleSideNavMock} push={pushMock} />, rrcMock.get());

    wrapper.find('div.sidenav-body>Link>button').at(3).simulate('click');
    expect(toggleSideNavMock).toHaveBeenCalled();
    expect(pushMock).toHaveBeenCalledWith('/orders');
  });

  it('should close open SideNav when close icon is clicked', () => {
    const toggleSideNavMock = jest.fn();
    const wrapper = mount(<SideNav user={caterer} logout={jest.fn()} open active="orders" toggleSideNav={toggleSideNavMock} />, rrcMock.get());

    wrapper.find('div.d-none-md>Link>button').simulate('click');
    expect(toggleSideNavMock).toHaveBeenCalled();
  });
});
