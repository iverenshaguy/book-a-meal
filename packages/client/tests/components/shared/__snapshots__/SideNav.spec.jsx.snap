// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should render SideNav component correctly 1`] = `
<div
  className="sidenav null"
  id="sidenav"
>
  <div
    className="sidenav-content"
  >
    <div
      className="sidenav-header"
    >
      <CloseIcon
        btnID={null}
        clickHandler={[Function]}
        divClass="d-none-md"
      />
      <div
        className="sidenav-title"
      >
        <h3
          className="d-flex-md"
        >
          <Link
            href="/"
            replace={false}
            to="/"
          >
            BOOK-A-MEAL
          </Link>
        </h3>
        <div
          className="username-circle"
        >
          <p>
            F
          </p>
        </div>
        <p>
          Food Circle
        </p>
        <span>
          <Link
            className={null}
            clickHandler={[MockFunction]}
            dataTip={null}
            id={null}
          >
            Logout
          </Link>
        </span>
      </div>
    </div>
    <div
      className="sidenav-body"
    >
      <Link
        className="menu-item active"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Dashboard
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Meals
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Menu
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Orders
      </Link>
    </div>
  </div>
</div>
`;

exports[`SideNav should render correctly when meals tab is active 1`] = `
<div
  className="sidenav null"
  id="sidenav"
>
  <div
    className="sidenav-content"
  >
    <div
      className="sidenav-header"
    >
      <CloseIcon
        btnID={null}
        clickHandler={[Function]}
        divClass="d-none-md"
      />
      <div
        className="sidenav-title"
      >
        <h3
          className="d-flex-md"
        >
          <Link
            href="/"
            replace={false}
            to="/"
          >
            BOOK-A-MEAL
          </Link>
        </h3>
        <div
          className="username-circle"
        >
          <p>
            F
          </p>
        </div>
        <p>
          Food Circle
        </p>
        <span>
          <Link
            className={null}
            clickHandler={[MockFunction]}
            dataTip={null}
            id={null}
          >
            Logout
          </Link>
        </span>
      </div>
    </div>
    <div
      className="sidenav-body"
    >
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Dashboard
      </Link>
      <Link
        className="menu-item active"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Meals
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Menu
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Orders
      </Link>
    </div>
  </div>
</div>
`;

exports[`SideNav should render correctly when menu tab is active 1`] = `
<div
  className="sidenav null"
  id="sidenav"
>
  <div
    className="sidenav-content"
  >
    <div
      className="sidenav-header"
    >
      <CloseIcon
        btnID={null}
        clickHandler={[Function]}
        divClass="d-none-md"
      />
      <div
        className="sidenav-title"
      >
        <h3
          className="d-flex-md"
        >
          <Link
            href="/"
            replace={false}
            to="/"
          >
            BOOK-A-MEAL
          </Link>
        </h3>
        <div
          className="username-circle"
        >
          <p>
            F
          </p>
        </div>
        <p>
          Food Circle
        </p>
        <span>
          <Link
            className={null}
            clickHandler={[MockFunction]}
            dataTip={null}
            id={null}
          >
            Logout
          </Link>
        </span>
      </div>
    </div>
    <div
      className="sidenav-body"
    >
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Dashboard
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Meals
      </Link>
      <Link
        className="menu-item active"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Menu
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Orders
      </Link>
    </div>
  </div>
</div>
`;

exports[`SideNav should render correctly when orders tab is active 1`] = `
<div
  className="sidenav null"
  id="sidenav"
>
  <div
    className="sidenav-content"
  >
    <div
      className="sidenav-header"
    >
      <CloseIcon
        btnID={null}
        clickHandler={[Function]}
        divClass="d-none-md"
      />
      <div
        className="sidenav-title"
      >
        <h3
          className="d-flex-md"
        >
          <Link
            href="/"
            replace={false}
            to="/"
          >
            BOOK-A-MEAL
          </Link>
        </h3>
        <div
          className="username-circle"
        >
          <p>
            F
          </p>
        </div>
        <p>
          Food Circle
        </p>
        <span>
          <Link
            className={null}
            clickHandler={[MockFunction]}
            dataTip={null}
            id={null}
          >
            Logout
          </Link>
        </span>
      </div>
    </div>
    <div
      className="sidenav-body"
    >
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Dashboard
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Meals
      </Link>
      <Link
        className="menu-item false"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Menu
      </Link>
      <Link
        className="menu-item active"
        clickHandler={[Function]}
        dataTip={null}
        id={null}
      >
        Orders
      </Link>
    </div>
  </div>
</div>
`;
